### PR TITLE
Adopt repositories

### PR DESCRIPTION
**Pull request by zeripath**
This PR is a rewrite of #12409 to remove the adopt/overwrite option as per @6543 and @lunny.

Prior to this PR Gitea would delete any repository files if they are
present during creation or migration. This can in certain circumstances
lead to data-loss and is slightly unpleasant.

This PR provides a mechanism for Gitea to adopt repositories or delete unadopted repositories by creating a new page
`/admin/repos/unadopted` which allow the administrator to search for unadopted repositories.

If normal users are allowed to adopt and/or delete unadopted repositories these will appear on `/user/settings/repos`

There is an API for the (administrator user only) `/api/v1/admin/unadopted` to list unadopted and POSTing to `/api/v1/admin/unadopted/:username/:reponame` will adopt whilst DELETE will delete them.

Closes #12409 

Signed-off-by: Andrew Thornton art27@cantab.net

### CODE REVIEW

File: routers/admin/repos.go [hunk 1]
```diff
@@ -50,3 +58,213 @@ func DeleteRepo(ctx *context.Context) {
...
+					break repoLoop
+				}
+			}
+			if !found {
+				if count >= start && count < end {
+					repoNames = append(repoNames, fmt.Sprintf("%s/%s", ctxUser.Name, name))
+				}
+				count++
+			}
+			if count >= end {
```
Comment (zeripath) [inline routers/admin/repos.go]: Oops. This needs removing.

File: routers/admin/repos.go [hunk 2]
```diff
@@ -50,3 +58,213 @@ func DeleteRepo(ctx *context.Context) {
...
+	root := filepath.Join(setting.RepoRootPath)
+	if err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() || path == root {
+			return nil
+		}
+
+		if strings.IndexRune(path[len(root)+1:], filepath.Separator) < 0 {
```
Comment (zeripath) [inline routers/admin/repos.go]: Golangci-lint prefers containsrune

File: routers/admin/repos.go [hunk 3]
```diff
@@ -50,3 +58,213 @@ func DeleteRepo(ctx *context.Context) {
...
+			return
+		}
+	} else if action == "delete" {
+		if err := repository.DeleteUnadoptedRepository(ctx.User, ctxUser, dirSplit[1]); err != nil {
+			ctx.ServerError("repository.AdoptRepository", err)
+			return
+		}
+	}
+	ctx.Redirect(setting.AppSubURL + "/admin/repos/unadopted")
+	return
```
Comment (zeripath) [inline routers/admin/repos.go]: And golangci-lint does not like this either

File: routers/user/setting/adopt.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+		}
+	} else if action == "delete" && allowDelete {
+		if err := repository.DeleteUnadoptedRepository(ctxUser, ctxUser, dir); err != nil {
+			ctx.ServerError("repository.AdoptRepository", err)
+			return
+		}
+	}
+
+	ctx.Redirect(setting.AppSubURL + "/user/settings/repos")
+	return
```
Comment (zeripath) [inline routers/user/setting/adopt.go]: Or this

Comment (zeripath) [reply routers/user/setting/adopt.go]:
```suggestion
```

File: routers/admin/repos.go (hunk 2 above)
Comment (zeripath) [reply routers/admin/repos.go]:
```suggestion
		if strings.ContainsRune(path[len(root)+1:], filepath.Separator) {
```

**Comment by 6543** [comment]
@zeripath one thing: the addoption task has to ensure to not set a not existing default branch

**Review by 6543 — APPROVED** [review]
👍

**Review by 6543 — APPROVED** [review]

File: modules/repository/create.go [hunk 5]
```diff
@@ -44,39 +46,77 @@ func CreateRepository(doer, u *models.User, opts models.CreateRepoOptions) (_ *m
...
+
+		repoPath := models.RepoPath(u.Name, repo.Name)
+		if com.IsExist(repoPath) {
+			// repo already exists - We have two or three options.
+			// 1. We fail stating that the directory exists
+			// 2. We create the db repository to go with this data and adopt the git repo
+			// 3. We delete it and start afresh
+			//
+			// Previously Gitea would just delete and start afresh - this was naughty.
+			// if opts.AdoptPreExisting {
```
Comment (lafriks) [inline modules/repository/create.go]: This commented out code should be deleted

**Review by lafriks — COMMENTED** [review]
Small nit about unneeded commented out code but otherwise lg-tm

File: modules/repository/create.go (hunk 5 above)
Comment (zeripath) [reply modules/repository/create.go]: Sorry missed that.

**Review by lafriks — APPROVED** [review]

**Pull request merged**